contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+            bscValidatorSet[i] = IBSCValidatorSet.Validator({
+                consensusAddress: validators[i],
+                feeAddress: payable(address(0)),
+                BBCFeeAddress: address(0),
+                votingPower: votingPowers[i],
+                jailed: false,
+                incoming: 0
+            });
+        }
+
+        // update validator set in BSCValidatorSet contract
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).updateValidatorSetV2(bscValidatorSet, voteAddrs);
+    }
+
+    function downtimeSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(downtimeSlashAmount);
+        _jailValidator(valInfo);
+
+        uint256 jailUntil = block.timestamp + downtimeJailTime;
+        if (valInfo.jailUntil < jailUntil) {
+            valInfo.jailUntil = jailUntil;
+        }
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, uint248(block.number), SlashType.DownTime);
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    function maliciousVoteSlash(bytes calldata _voteAddr, uint256 height) external onlySlash {
+        address operatorAddress = _voteToOperator[_voteAddr];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(doubleSignSlashAmount);
+        _jailValidator(valInfo);
+
+        SlashRecord storage record = _updateSlashRecord(operatorAddress, slashAmount, height, SlashType.MaliciousVote);
+        if (valInfo.jailUntil < record.jailUntil) {
+            valInfo.jailUntil = record.jailUntil;
+        }
+
+        emit ValidatorSlashed(
+            operatorAddress, record.jailUntil, record.slashAmount, record.slashHeight, SlashType.MaliciousVote
+        );
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    function doubleSignSlash(address consensusAddress, uint256 height, uint256 evidenceTime) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+        require(evidenceTime + maxEvidenceAge >= block.timestamp, "EVIDENCE_TOO_OLD");
```

Comment:
this check should not happen here, the check should be done before invoke `doubleSignSlash`, only the evidence is valid, we can go to the code here.

Reply:
Done